src/crypto/CHIPCryptoPAL.h

```diff
@@ -214,6 +262,426 @@ CHIP_ERROR add_entropy_source(entropy_source fn_source, void * p_source, size_t
+                             size_t peer_identity_len, const unsigned char * w0in, size_t w0in_len, const unsigned char * Lin,
+                             size_t Lin_len);
+
+    /**
+     * @brief Start the Spake2+ process as a prover (i.e. a commisioner).
+     *
+     * @param my_identity       The prover identity. May be NULL if identities are not established.
+     * @param my_identity_len   The prover identity length.
+     * @param peer_identity     The peer identity. May be NULL if identities are not established.
+     * @param peer_identity_len The peer identity length.
+     * @param w0in              The input w0 (an output from the PBKDF).
+     * @param w0in_len          The input w0 length.
+     * @param w1in              The input w1 (an output from the PBKDF).
+     * @param w1in_len          The input w1 length.
+     *
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    CHIP_ERROR BeginProver(const unsigned char * my_identity, size_t my_identity_len, const unsigned char * peer_identity,
+                           size_t peer_identity_len, const unsigned char * w0in, size_t w0in_len, const unsigned char * w1in,
+                           size_t w1in_len);
+
+    /**
+     * @brief Compute the first round of the protocol.
+     *
+     * @param out     The output first round Spake2+ contribution.
+     * @param out_len The output first round Spake2+ contribution length.
+     *
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    CHIP_ERROR ComputeRoundOne(unsigned char * out, size_t * out_len);
+
+    /**
+     * @brief Compute the second round of the protocol.
+     *
+     * @param in      The peer first round Spake2+ contribution.
+     * @param in_len  The peer first round Spake2+ contribution length.
+     * @param out     The output second round Spake2+ contribution.
+     * @param out_len The output second round Spake2+ contribution length.
+     *
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    CHIP_ERROR ComputeRoundTwo(const unsigned char * in, size_t in_len, unsigned char * out, size_t * out_len);
+
+    /**
+     * @brief Confirm that each party computed the same keys.
+     *
+     * @param in     The peer second round Spake2+ contribution.
+     * @param in_len The peer second round Spake2+ contribution length.
+     *
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    CHIP_ERROR KeyConfirm(const unsigned char * in, size_t in_len);
+
+    /**
+     * @brief Return the shared secret.
+     *
+     * @param out     The output secret.
+     * @param out_len The output secret length.
+     *
+     * @return Returns a CHIP_ERROR on error, CHIP_NO_ERROR otherwise
+     **/
+    CHIP_ERROR GetKeys(unsigned char * out, size_t * out_len);
+
+    CHIP_SPAKE2P_ROLE role;
```

Comment:
do these members need to be public or could we make them a bit more restricted?

Reply:
Also, shouldn't the data fields be private?